exchange/accounts/balance.go

```diff
@@ -0,0 +1,169 @@
+package accounts
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+)
+
+var (
+	errBalanceCurrencyMismatch = errors.New("balance currency does not match update currency")
+	errOutOfSequence           = errors.New("out of sequence")
+	errUpdatedAtIsZero         = errors.New("updatedAt may not be zero")
+)
+
+// Balance contains an exchange currency balance
+type Balance struct {
+	Currency               currency.Code
+	Total                  float64
+	Hold                   float64
+	Free                   float64
+	AvailableWithoutBorrow float64
+	Borrowed               float64
+	UpdatedAt              time.Time
+}
+
+// Change defines incoming balance change on currency holdings
+type Change struct {
+	Account   string
+	AssetType asset.Item
+	Balance   Balance
+}
+
+// balance contains a balance with live updates
+type balance struct {
+	internal Balance
+	m        sync.RWMutex
+}
+
+// CurrencyBalances provides a map of currencies to balances
+type CurrencyBalances map[currency.Code]Balance
+
+// currencyBalances provides a map of currencies to balances
+type currencyBalances map[currency.Code]*balance
+
+// Set will set a currency balance, overwriting any previous Balance
+// currency may be a string or a currency.Code
+func (c *CurrencyBalances) Set(cAny any, b Balance) { //nolint:gocritic // hugeparam not relevant; we want to store a value so we'd deref anyway
+	var curr currency.Code
+	switch v := cAny.(type) {
+	case string:
+		curr = currency.NewCode(v)
+	case currency.Code:
+		curr = v
+	default:
+		panic(fmt.Errorf("%w: '%v'", currency.ErrCurrencyNotSupported, cAny))
+	}
+	b.Currency = curr
+	(*c)[curr] = b
+}
+
+// Add will add to a currency balance
+// currency may be a string or a currency.Code
+func (c *CurrencyBalances) Add(cAny any, b Balance) { //nolint:gocritic // hugeparam not relevant; we want to store a value so we'd deref anyway
```

Comment:
String case seems to only be used in Yobit, I would just forced `c currency.Code` and the burden is shifted to the caller. I would do that for Set as well. It just reduces complexity/panic potential.

Reply:
Agreed.
I'll also shift the other panic out to the callers.

Bit of a rabbit hole; At first I made all the callers use `currency.NewCode`, but actually they should be adjusting their data types, mostly.